[Ready for Review] MD5+SHA256 Hashes available in the file view page and through the api.

hmoco opened the pull request:
## Purpose
Add md5 and sha256 information for osfstorage files and when other addons provide it. Then, use that information on the file view page to view the hashes for every version and every file in osfstorage. Also allows future use of the hashes through the api. 
This accompanies waterbutler PR: https://github.com/CenterForOpenScience/waterbutler/pull/71, and both need to be merged for it to work.

## Changes
Add a hashes section to the revision view in file page:
![screen shot 2015-08-11 at 1 19 15 pm](https://cloud.githubusercontent.com/assets/6268982/9204612/243b6f9c-402c-11e5-9ba9-7cd451b8ac2a.png)

Also md5 and sha256 hash for any file is available through the API under ```metadata['extra']['hashes']```

and scroll bar added to revisions tab.

## Side Effects
Fixes the problem below in production by adding a scroll bar:
![screen shot 2015-07-14 at 10 20 18 am](https://cloud.githubusercontent.com/assets/6268982/8675454/0d94711a-2a12-11e5-9e46-250895fbc938.png)

Also, the copy button uses zeroClipboard, meaning it will only work if the browser has flash 11+ enabled. 
https://github.com/zeroclipboard/zeroclipboard/blob/master/docs/instructions.md#limitations
It also has known bugs with IE (due to being within a modal) and Opera.
https://github.com/zeroclipboard/zeroclipboard/blob/master/docs/instructions.md#known-conflicts-with-other-libraries 


## Potential Testing (QA)

Make sure the copy button works (it won’t for some IE versions)
Make sure only osfstorage files have the column with buttons
Make sure each version has different hashes
Upload the same file to two different account/project, make sure the hashes match
Use a hash calculator to make sure the hash is correct. (Any website that allows you to upload a file and see its md5 or sha256 hash)

chrisseto commented on website/addons/osfstorage/utils.py:
```diff
@@ -49,6 +49,7 @@ def serialize_revision(node, record, version, index, anon=False):
         'index': index + 1,
         'date': version.date_created.isoformat(),
         'downloads': record.get_download_count(version=index),
+        'md5': md5
```
You can just pull this out of `version.metadata['hashes']['md5']`

hmoco replied:
I wasn't able to access anything within version :/, will give this a try.

chrisseto commented on website/static/js/filepage/revisions.js:
```diff
@@ -29,7 +30,10 @@ var FileRevisionsTable = {
+        self.md5 = [];
+        self.buttonToText = [];
 
+        model.hasMd5 = (self.file.provider === 'osfstorage') || (self.file.provider === 's3');
```
Be careful with S3 as the md5 is not always the md5 of the file.

hmoco replied:
Ok, we can have just osfstorage for now

hmoco commented:
@chrisseto This might be good to go.

hmoco commented on website/static/js/filepage/revisions.js:
```diff
@@ -110,6 +117,35 @@ var FileRevisionsTable = {
+                            }
+                        }
+                    }, m('td', self.sha256[revision.displayVersion] || 'SHA256') )
+                ) : false,
```
@chrisseto Please look at this mess

chrisseto commented:
Another thought for the UI which may make things a touch more simple.
Use a bootstrap label element containing the hash type. And then on mouseover display the hash's value.
Pinging @caneruguz for his input here.

hmoco commented:
After talking to Jeff, this is the new UI for the buttons:
![gif3](https://cloud.githubusercontent.com/assets/6268982/8651940/d0e4b388-2949-11e5-83d0-b18a8f758f6d.gif)

hmoco commented:
@sloria @chrisseto I believe this is good for code review. Reminder: This can't  be merged without also merging https://github.com/CenterForOpenScience/waterbutler/pull/58

lyndsysimon commented on website/static/js/filepage/revisions.js:
```diff
@@ -86,12 +90,31 @@ var FileRevisionsTable = {
+                    '<div class="col-md-12">' +
+                    '<div class="control-group">' +'<p class="col-xs-10">' + hash + '</p>' +
+                    '<a class="btn-sm btn-primary col-xs-2" style="align: right; text-align: center; ' +
+                    'padding: 5px; width: 80px" id="copyBtn" data-clipboard-text="' +
```
Can we do without this hardcoded width? ("`80px`").

Also, these buttons are touching, but are not part of a button bar. They should either be separated by an appropriate margin, or they should be joined into a bar. I defer to @caneruguz as to what it should look like.

Finally, the height of these buttons is inconsistent with the height of the download button to the right

hmoco replied:
The hardcoded width was the best way I found to make the 'copy' button behave as expected, but I can look into better solutions. I'm assuming you are talking about the hashes button here though, not the copy button.

lyndsysimon commented:
The clipping problem should be addressed using Bootstrap's [responsive utlity classes](http://getbootstrap.com/css/#responsive-utilities).

lyndsysimon commented:
First pass complete - no bugs found, but please take a look at the UI issues I've noted.

sloria commented on website/static/js/filepage/revisions.js:
```diff
@@ -86,12 +90,31 @@ var FileRevisionsTable = {
+            });
+            showModal.on('show.bs.modal', function() {
+                var $copyBtn = $('#copyBtn');
+                new makeClient($copyBtn);
```
`makeClient` is not a constructor; it is a factory for ZeroClipboard clients, so no need to `new` here.

sloria replied:
This will create a new client every time one of the modals is shown, but the clients aren't destroyed when the modal is closed. Either (1) don't create a new client on every modal shown or (2) make sure to clean up the clients after closing the modal.

sloria commented on website/static/js/filepage/revisions.js:
```diff
@@ -2,6 +2,8 @@ var m = require('mithril');
 var $osf = require('js/osfHelpers');
 var waterbutler = require('js/waterbutler');
+var bootbox = require('bootbox');
+var makeClient = require('js/clipboard.js');
```
Omit the `.js` extension.

sloria commented on website/static/js/filepage/revisions.js:
```diff
@@ -2,6 +2,8 @@ var m = require('mithril');
 var $ = require('jquery');
 var $osf = require('js/osfHelpers');
 var waterbutler = require('js/waterbutler');
+var bootbox = require('bootbox');
```
Group imports according to our style guidelines: https://cosdev.readthedocs.org/en/latest/style_guides/javascript.html#commonjs-modules

sloria commented on website/static/js/filepage/revisions.js:
```diff
@@ -30,6 +33,7 @@ var FileRevisionsTable = {
         self.enableEditing = enableEditing;
         self.baseUrl = (window.location.href).split('?')[0];
 
+        model.hasHashes = self.file.provider === 'osfstorage';
```
Rather than checking against addon name, can you instead check against the existence of the `md5` and `sha256` metadata? Keep in mind we may support hashes for other addons in the future.

sloria commented:
Added a few drive-by comments, but haven't done full review; @hmoco pass this back to Lyndsy when you are done addressing the above review.

hmoco commented:
@lyndsysimon, I believe I have addressed Steve's comments. For yours, I have changed the button size and made them into a button group. However, I have not found a better way to represent the copy button, and the responsive classes from bootstrap (hidden, visible) seem to have to effect on mithril table objects.

hmoco commented:
Do not pull, current;y broken

hmoco commented:
Not sure if fix is perfect, but it works now.

sloria commented:
Did a quick pass; once the UI is updated with the changes we discussed, I'll look at this more closely.

caneruguz commented:
@hmoco You are only showing Revisions table in a very wide width. Can you try out narrow widths and make sure your UI does not break in these cases?

hmoco commented:
@caneruguz Thanks to the scrolling, and the fixed widths on the hash columns, it does not break

sloria commented:
Looks good on first glance. Will look at this more closely once https://github.com/CenterForOpenScience/osf.io/pull/4111 is merged

sloria commented on website/static/js/filepage/revisions.js:
```diff
@@ -120,6 +126,20 @@ var FileRevisionsTable = {
+                model.hasHashes ? m('td',
+                    m('div.input-group[style="width: 180px"]',
+                        [
+                            m('span.input-group-btn', m('button#copyBtnMd5'+ revision.displayVersion +'.btn.btn-default.btn-sm[type="button"][data-clipboard-text="'+revision.extra.hashes.md5 + '"]', {config: new makeClient($('#copyBtnMd5'+ revision.displayVersion))}, m('.fa.fa-copy'))),
```
The `config` param should be a function; you are passing it a ZeroClipboard client object. Notice that `makeClient` is not a constructor function. It is a factory, so the `new` keyword is incorrect usage.

Also, `makeClient` takes a native DOM element object (not a jQuery object). Conveniently, the `config` function receives the DOM element, so you can pass that to `makeClient`. This also means you can remove the unnecessary CSS IDs.

sloria commented:
Pass finished. :key: Take a look at the `makeClient` function to understand its correct usage, and modify your code accordingly.

sloria commented:
Also, can you get this up to date with the `release/0.48.0` branch and reconcile these changes with your "Revisions view" changes?
